lib/vrl/stdlib/src/parse_host.rs

```diff
@@ -0,0 +1,192 @@
+                keyword: "private_domains",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "parse domain name",
+                source: r#"parse_host!("www.example.com")"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "example",
+                        "subdomain": "www",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+            Example {
+                title: "parse URL",
+                source: r#"parse_host!("https://www.example.com/s?q=foobar")"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "example",
+                        "subdomain": "www",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+            Example {
+                title: "ignore private domain",
+                source: r#"parse_host!("internal.us-east-1.elb.amazonaws.com", private_domains: false)"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "amazonaws",
+                        "subdomain": "internal.us-east-1.elb",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+        ]
+    }
+
+    fn compile(&self, _state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        let private_domains = arguments
+            .optional("private_domains")
+            .unwrap_or_else(|| expr!(true));
+
+        let options = TldOption {
+            cache_path: Some("/tmp/.tld_cache".to_string()),
+            ..Default::default()
+        };
+        let private_options = TldOption {
+            cache_path: Some("/tmp/.private_tld_cache".to_string()),
+            private_domains: true,
+            ..Default::default()
+        };
+
+        Ok(Box::new(ParseHostFn {
+            value,
+            options,
+            private_domains,
+            private_options,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseHostFn {
+    value: Box<dyn Expression>,
+    options: TldOption,
+    private_domains: Box<dyn Expression>,
+    private_options: TldOption,
+}
+
+impl Expression for ParseHostFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let string = value.try_bytes_utf8_lossy()?;
+
+        let private_domains = self.private_domains.resolve(ctx)?.try_boolean()?;
+        let options = self.options.clone();
+        let private_options = self.private_options.clone();
+
+        let ext = if private_domains {
+            TldExtractor::new(private_options)
+        } else {
+            TldExtractor::new(options)
+        };
```

Comment:
It's actually the `TldExtractor::new` call that does the file-system lookup, which would be expensive to do for each individual event.

We'd want `TldExtractor` to be stored in `ParseHostFn`, but I realize that that would require `TldExtractor` to impl `Clone`, which it doesn't.

So, that would suggest there isn't an easy way to make this as performant as we want.

We could at least benchmark the existing implementation, to see if we're happy shipping it as-is, or not.

Reply:
> We'd want `TldExtractor` to be stored in `ParseHostFn`, but I realize that that would require `TldExtractor` to impl `Clone`, which it doesn't.

We can resolve this by wrapping in an `Arc`.

Reply:
I could look to PR that as well 🤷